text/0025-confidential-token-ids.md

```diff
@@ -0,0 +1,220 @@
+
+* To compute the 4 byte `masked_token_id` from the `token_id`, we hash the TxOut shared secret, with a prefix, XORed with the `token_id`, and take the little endian representation of those bytes.
+
+    ```
+    masked_token_id = (token_id ^ Blake2B(AMOUNT_TOKEN_ID_DOMAIN_TAG | shared_secret)).to_le()
+    ```
+
+* To recover the `token_id` from the `masked_token_id`, we reverse the process (if the bytes field is length = 4, otherwise `token_id = 0`).
+
+    ```
+    unmasked_token_id = masked_token_id ^ Blake2b(AMOUNT_TOKEN_ID_DOMAIN_TAG | shared_secret)
+    ```
+
+### Pedersen Generators
+
+Previously, we used the same curve base point for all Pedersen generators to Amounts. After this change, we will need a curve base point, unique for each `token_id`, from which the TxOut's commitment is generated with the value of the `token_id` in the TxOut.
+
+The requirements for the Pedersen generators are:
+
+* The generator used for `MOB = token_id(0)` is the same as before, so `H_0` is a known value.
+
+* The set of all generators used for `token_ids` should be cryptographically "orthogonal", meaning that, it should be intractable for anyone to find a linear relationship between them.
+
+* The function mapping a `token_id`, `i`, to the generator `H_i` must be implemented in constant-time.
+
+The established value for `H_0` in mobilecoin is to take the `RISTRETTO_BASEPOINT` bytes, hash them using Blake2b, and then hash this to the ristretto curve using the elligator map exposed by curve25519-dalek.
+
+We propose to compute `H_i` by converting `i` to little endian bytes, and XOR'ing these bytes over the first four bytes of the `RISTRETTO_BASEPOINT` bytes before hashing them with Blake2b, because this is a simple way to meet these requirements.
+
+For example, an implementation to obtain a generator for `token_id = i` could look like:
+
+```
+B = RistrettoPoint::from_hash(Blake2b(HASH_TO_POINT_DOMAIN_TAG | RISTRETTO_BASEPOINT ^ token_id))
+B_blinding = RISTRETTO_BASEPOINT
+```
+
+After this change, `range_proofs`, `rct_bulletproofs`, and `ring_signature/mlsags` will be relative to a Pedersen generator, and it is not possible to construct a range proof relative to another generator if those generators are orthogonal. Thus, it is guaranteed that all transaction inputs and outputs are using the same `token_ids`. This is implied by the homomorphic encryption property of [Pedersen Commitments](https://www.cs.cornell.edu/courses/cs754/2001fa/129.PDF), namely that addition on the commitments preserves the additive relationship of the pre-committed values, as long as they are using the same group generator. `H_i` is revealed to the verifier, who must show that the sum of inputs does not overflow, or `sum(outputs) = a * G + b * H_i` for some TxOut-specific values of `a` and `b`. 
```

Comment:
Modifying range proof structures to accommodate this proposal is intrusive, prevents efficient batch-verification which directly affects the upper bound of tx throughput, and may impose a burden on/barrier to future proof upgrades.

Reply:
That's a good point, and a reason to do it the way the Spats proposal works.

I think right now we don't have any of the batch verification stuff in place, so we'll have to think seriously about this. I don't know if this is actually the bottleneck right now in transaction validation.